### PR TITLE
[NPU] add AMP O1 support

### PR DESCRIPTION
**Pull request by windstamp**
### PR types

Bug fixes
### PR changes

Others
### Describe

[NPU] add AMP O1 support

### CODE REVIEW

File: python/paddle/fluid/dygraph/amp/auto_cast.py [hunk 1]
```diff
@@ -274,9 +274,10 @@ def amp_guard(enable=True,
     # NOTE: Now, amp only support gpu for float16 and bfloat16, xpu for float16.
```
Comment (zhangbo9674) [inline python/paddle/fluid/dygraph/amp/auto_cast.py]: 补充完善一下NOTE?

File: python/paddle/fluid/dygraph/amp/auto_cast.py [hunk 2]
```diff
@@ -274,9 +274,10 @@ def amp_guard(enable=True,
...
     if enable and not (tracer._expected_place.is_gpu_place() or
-                       tracer._expected_place.is_xpu_place()):
+                       tracer._expected_place.is_xpu_place() or
+                       tracer._expected_place.is_npu_place()):
         warnings.warn(
-            'amp_guard can only be enabled on CUDAPlace and XPUPlace, current place is %s, so it makes no effect.'
+            'amp_guard can only be enabled on CUDAPlace, XPUPlace, and NPUPlace, current place is %s, so it makes no effect.'
             % tracer._expected_place)
         enable = False
     # For xpu:
```
Comment (zhangbo9674) [inline python/paddle/fluid/dygraph/amp/auto_cast.py:287]: 类似于这里的 For xpu，做下判断，npu目前只支持float16

File: python/paddle/fluid/dygraph/amp/auto_cast.py (hunk 1 above)
Comment (windstamp) [reply python/paddle/fluid/dygraph/amp/auto_cast.py]: Done, thanks.

File: python/paddle/fluid/dygraph/amp/auto_cast.py (hunk 2 above)
Comment (windstamp) [reply python/paddle/fluid/dygraph/amp/auto_cast.py:287]: Done, thanks.

**Review by zhangbo9674 — APPROVED** [review]
LGTM

**Review by zhiqiu — APPROVED** [review]
LGTM

**Pull request merged**